### PR TITLE
tracking: make opencv_dnn dependancy optional

### LINKED ISSUE #1030: 'opencv2/datasets/track_alov.hpp' file not found
detail in https://github.com/opencv/opencv/issues/8303

### LINKED ISSUE #1036: Module opencv_tracking disabled because opencv_dnn dependency can't be resolved!
I run opencv-3.2.0/platforms/ios/build_framework.py with command:
./build_framework.py --contrib ../../opencv_contrib ios

But "opencv2.framework" not generate tracking module, I find follow log:
-- Module opencv_sfm disabled because the following dependencies are not found: Glog/Gflags
-- Module opencv_tracking disabled because opencv_dnn dependency can't be resolved!

I want to use the KCF and TLD algorithm in opencv_tracking, how to use it on ios ?

### PR DESCRIPTION
**Pull request by berak**
resolves #1036
resolves #1030
resolves #1046

### This pullrequest changes

add a few HAVE_OPENCV_DNN conditionals, based on opencv_modules.hpp

### CODE REVIEW

File: modules/tracking/src/gtrUtils.cpp [hunk 1]
```diff
@@ -144,3 +147,5 @@ vector <TrainingSample> gatherFrameSamples(Mat prevFrame, Mat currFrame, Rect2f
 
 }
 }
+
+#endif // HAVE_OPENCV_DNN
```
Comment (alalek) [inline modules/tracking/src/gtrUtils.cpp]:
```
#else
    ... make CV_Error() stubs for public API methods, like `TrackerGOTURN::createTracker()` and others
#endif
```

**Comment by berak** [comment]
conditionals applied more locally, per function(implementation) now.

File: modules/tracking/src/gtrTracker.cpp [hunk 2]
```diff
@@ -54,7 +54,13 @@ void TrackerGOTURN::Params::write(cv::FileStorage& /*fs*/) const {}
 
 Ptr<TrackerGOTURN> TrackerGOTURN::createTracker(const TrackerGOTURN::Params &parameters)
 {
+#ifdef HAVE_OPENCV_DNN
     return Ptr<gtr::TrackerGOTURNImpl>(new gtr::TrackerGOTURNImpl(parameters));
+#else
+    (void)(parameters);
+    CV_Error(cv::Error::StsNotImplemented , "to use GOTURN, the tracking module needs to be built with opencv_dnn !");
+    return Ptr<TrackerGOTURN>();
```
Comment (alalek) [inline modules/tracking/src/gtrTracker.cpp]: Is `TrackerGOTURN` abstract class? If so, it is better to use `CV_ErrorNoReturn` without "return" statement.

File: modules/tracking/src/gtrTracker.hpp [hunk 3]
```diff
@@ -67,7 +72,9 @@ class TrackerGOTURNImpl : public TrackerGOTURN
 
     TrackerGOTURN::Params params;
 
+#ifdef HAVE_OPENCV_DNN
     dnn::Net net;
+#endif
```
Comment (alalek) [inline modules/tracking/src/gtrTracker.hpp]: I believe we can hide `TrackerGOTURNImpl` completely.

File: modules/tracking/src/gtrTracker.cpp [hunk 4]
```diff
@@ -74,6 +79,7 @@ class TrackerGOTURNModel : public TrackerModel{
     void modelUpdateImpl(){}
 };
 
+#ifdef HAVE_OPENCV_DNN
```
Comment (alalek) [inline modules/tracking/src/gtrTracker.cpp]: Looks like `TrackerGOTURNModel` (from line 65/68) is not needed too.

**Comment by alalek** [comment]
@berak Could you please fix tracker_dataset.cpp file: need to add "#include <opencv2/core.hpp>" into "else" branch too:
```
opencv_contrib/modules/tracking/samples/tracker_dataset.cpp:238:11: error: ‘cv’ has not been declared
  CV_Error(cv::Error::StsNotImplemented , "this sample needs to be built with opencv_datasets !");
```

**Comment by berak** [comment]
sure, will do !

**Comment by alalek** [comment]
Well done. Thank you!
:+1:

**Pull request merged**